File: 2021/0108_sortedArrayToBST/0817/sortedArrayToBST.py
class Solution:
    def sortedArrayToBST(self, nums: List[int]):
        return self.makeTree(nums, 0, len(nums) - 1)

    def makeTree(self, nums, leftIndex, rightIndex):
        if leftIndex > rightIndex: return None
        midIndex = (leftIndex + rightIndex) // 2
        root = TreeNode(nums[midIndex])
        root.left = self.makeTree(nums, leftIndex, midIndex - 1)
        root.right = self.makeTree(nums, midIndex + 1, rightIndex)
        return root

File: 2021/0108_sortedArrayToBST/0817/sortedArrayToBST.rs
impl Solution {
    pub fn sortedArrayToBST(nums: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        Self::build_bst(&nums) 
    } 

    fn build_bst(nums: &[i32]) -> Option<Rc<RefCell<TreeNode>>> {
        if nums.is_empty() { return None }
        let mid = nums.len() / 2;
        Some(Rc::new(RefCell::new(TreeNode {
            val: nums[mid],
            left: Self::build_bst(&nums[..mid]),
            right: Self::build_bst(&nums[mid + 1..]),
        })))
    }
}
